pallets/stateful-storage/src/lib.rs

```diff
@@ -290,4 +336,33 @@ pub mod pallet {
 	}
 }
 
-impl<T: Config> Pallet<T> {}
+impl<T: Config> Pallet<T> {
+	fn check_origin(origin: OriginFor<T>) -> Result<MessageSourceId, DispatchError> {
+		let provider_key = ensure_signed(origin)?;
+		Ok(T::MsaInfoProvider::ensure_valid_msa_key(&provider_key)
+			.map_err(|_| Error::<T>::InvalidMessageSourceAccount)?)
+	}
+
+	fn check_schema_and_grants(
+		provider_msa_id: MessageSourceId,
+		state_owner_msa_id: MessageSourceId,
+		schema_id: SchemaId,
+		payload_location: PayloadLocation,
+	) -> DispatchResult {
+		let schema =
+			T::SchemaProvider::get_schema_by_id(schema_id).ok_or(Error::<T>::InvalidSchemaId)?;
+		ensure!(
+			schema.payload_location == payload_location,
+			Error::<T>::SchemaPayloadLocationMismatch
+		);
+
+		let current_block = frame_system::Pallet::<T>::block_number();
```

Comment:
I wonder if removing this call is ok:
```
let schema =
			T::SchemaProvider::get_schema_by_id(schema_id).ok_or(Error::<T>::InvalidSchemaId)?;
		ensure!(
			schema.payload_location == payload_location,
			Error::<T>::SchemaPayloadLocationMismatch
		);
```

It might be that we do not need it if you can grant that schema implies that the grant exists?

```
T::SchemaGrantValidator::ensure_valid_schema_grant(
			ProviderId(provider_msa_id),
			DelegatorId(state_owner_msa_id),
			schema_id,
			current_block,
		)
		```

Reply:
Problem is we need to verify the payload location matches. That information only exists on the schema itself, we can't get or infer it from the grant.

Reply:
ahhh, yeah your right.